vk-video/src/parser.rs

```diff
@@ -77,99 +162,33 @@ pub enum DecoderInstruction {
     Pps(PicParameterSet),
 }
 
-#[derive(Debug, thiserror::Error)]
-pub enum ParserError {
-    #[error(transparent)]
-    ReferenceManagementError(#[from] ReferenceManagementError),
-
-    #[error("Bitstreams that allow gaps in frame_num are not supported")]
-    GapsInFrameNumNotSupported,
-
-    #[error("Streams containing fields instead of frames are not supported")]
-    FieldsNotSupported,
-
-    #[error("Error while parsing a NAL header: {0:?}")]
-    NalHeaderParseError(h264_reader::nal::NalHeaderError),
-
-    #[error("Error while parsing SPS: {0:?}")]
-    SpsParseError(h264_reader::nal::sps::SpsError),
-
-    #[error("Error while parsing PPS: {0:?}")]
-    PpsParseError(h264_reader::nal::pps::PpsError),
-
-    #[error("Error while parsing a slice: {0:?}")]
-    SliceParseError(h264_reader::nal::slice::SliceHeaderError),
-}
-
-pub struct Parser {
-    reader: AnnexBReader<NalAccumulator<NalReceiver>>,
-    reference_ctx: ReferenceContext,
-    au_splitter: AUSplitter,
-    receiver: mpsc::Receiver<Result<ParsedNalu, ParserError>>,
-    nalu_splitter: NALUSplitter,
-}
-
-impl Parser {
-    pub fn new(missed_frame_handling: MissedFrameHandling) -> Self {
-        let (tx, rx) = mpsc::channel();
-
-        Parser {
-            reader: AnnexBReader::accumulate(NalReceiver::new(tx)),
-            reference_ctx: ReferenceContext::new(missed_frame_handling),
-            au_splitter: AUSplitter::default(),
-            receiver: rx,
-            nalu_splitter: NALUSplitter::default(),
-        }
-    }
-
-    pub fn parse(
-        &mut self,
-        bytes: &[u8],
-        pts: Option<u64>,
-    ) -> Result<Vec<DecoderInstruction>, ParserError> {
-        let nalus = self.nalu_splitter.push(bytes, pts);
-        let nalus = nalus
-            .into_iter()
-            .map(|(nalu, pts)| {
-                self.reader.push(&nalu);
-                (self.receiver.try_recv().unwrap(), pts)
-            })
-            .collect::<Vec<_>>();
-
-        let mut instructions = Vec::new();
-        for (nalu, pts) in nalus {
-            let nalu = nalu?;
-
-            let Some(nalus) = self.au_splitter.put_nalu(nalu, pts) else {
-                continue;
-            };
-
-            let mut slices = Vec::new();
-            for (nalu, pts) in nalus {
-                match nalu {
-                    ParsedNalu::Sps(seq_parameter_set) => {
-                        instructions.push(DecoderInstruction::Sps(seq_parameter_set))
-                    }
-                    ParsedNalu::Pps(pic_parameter_set) => {
-                        instructions.push(DecoderInstruction::Pps(pic_parameter_set))
-                    }
-                    ParsedNalu::Slice(slice) => {
-                        slices.push((slice, pts));
-                    }
-
-                    ParsedNalu::Other(_) => {}
+pub(crate) fn extract_decoder_instructions(
```

Comment:
How about a fun name like `compile_to_decoder_instructions`? I always imagined decoding h264 as writing an emulator for an obscure game console, where videos are the game ROMs and this step is compilation of the rom data into a VM bytecode. I think it's quite fun and that the analogy works surprisingly well.

Then in the future we can think of various codecs as compiler front-ends, which they essentially will be.